api/server/runner_fninvoke.go

```diff
@@ -66,31 +78,35 @@ func (s *Server) handleFnInvokeCall2(c *gin.Context) error {
 }
 
 func (s *Server) ServeFnInvoke(c *gin.Context, app *models.App, fn *models.Fn) error {
-	// TODO: we should combine this logic with trigger, which just wraps this block with some headers wizardry
 	// TODO: we should get rid of the buffers, and stream back (saves memory (+splice), faster (splice), allows streaming, don't have to cap resp size)
+	//TODO: Remove this pool, it makes concurrency bugs easy to create, but hard to discover.
```

Comment:
the easy thing to do is make sure that submit returned successfully and the buffer was drained, which will probably be >99% of buffers... we need the pool tho for memory pressure, if we're to keep the buffers it's basically mandatory (fragmentation, gc pressure, etc...) since these will be the largest chunks of memory for this program. eg check for success (sorry, was lazy adding this the first time around, but dispatch with the pipes actually protected us ok):

```
defer func() {
  if buf.Len() == 0 && submitErr != nil {
    bufPool.Put(buf)
  }
}
```

Reply:
yeah, but ideally we wouldn't be storing responses in memory, they should be just streamed straight back to the client, as the other note notes.